docs/book/start/getting-data-to-user-space.md

```diff
@@ -0,0 +1,174 @@
+
+`PerfEventArray` is a collection of per-CPU circular buffers that enable the
+kernel to emit events (defined as custom structs) to user space. Each CPU has
+its own buffer, and the eBPF program emits an event to the buffer of the CPU
+it's currently running on. The events are out of order, meaning that they arrive
+in the user-space in a different order than they were created and sent from the
+eBPF program.
+
+To gather events from all CPUs, we are going to spawn a thread for each CPU to
+poll for the events and then iterate over them.
+
+The data structure we'll be using needs to hold an IPv4 address and a port.
+
+```rust linenums="1" title="xdp-perfbuf-custom-data-common/src/lib.rs"
+--8<-- "examples/xdp-perfbuf-custom-data/xdp-perfbuf-custom-data-common/src/lib.rs"
+```
+
+1. We implement the `aya::Pod` trait for our struct since it is Plain Old Data
+as can be safely converted to a byte slice and back.
+
+1. Events emitted with `PerfEventArray` are copied from kernel memory to user
+memory, therefore they must implement the `aya::Pod` trait (where `Pod` stands
+for "plain old data") which expresses that it's safe to convert them into a
+sequence of bytes.
+
+!!! tip "Alignment, padding and verifier errors"
+
+    At program load time, the eBPF verifier checks that all the memory used is
+    properly initialized. This can be a problem if - to ensure alignment - the
+    compiler inserts padding bytes between fields in your types.
+
+    **Example:**
+
+    ```rust
+    #[repr(C)]
+    struct SourceInfo {
+        source_port: u16,
+        source_ip: u32,
+    }
+
+    let source_port = ...;
+    let source_ip = ...;
+    let si = SourceInfo { source_port, source_ip };
+    ```
+
+    In the example above, the compiler will insert two extra bytes between the
+    struct fields `source_port` and `source_ip` to make sure that `source_ip` is
+    correctly aligned to a 4-byte address (assuming `mem::align_of::<u32>() ==
+    4`).  Since padding bytes are typically not initialized by the compiler,
+    this will result in the infamous `invalid indirect read from stack` verifier
+    error.
+
+    To avoid the error, you can either manually ensure that all the fields in
+    your types are correctly aligned (e.g. by explicitly adding padding or by
+    making field types larger to enforce alignment) or use `#[repr(packed)]`.
+    Since the latter comes with its own foot-guns and can perform less
+    efficiently, explicitly adding padding or tweaking alignment is recommended.
+
+    **Solution ensuring alignment using larger types:**
+
+    ```rust
+    #[repr(C)]
+    struct SourceInfo {
+        source_port: u32,
+        source_ip: u32,
+    }
+
+    let source_port = ...;
+    let source_ip = ...;
+    let si = SourceInfo { source_port, source_ip };
+    ```
+
+    **Solution with explicit padding:**
+
+    ```rust
+    #[repr(C)]
+    struct SourceInfo {
+        source_port: u16,
+        padding: u16,
```

Comment:
should this be `_padding`?

Reply:
I will also make all the other fields public.